### PR TITLE
Feature/general purpose table constraints

### PR DESCRIPTION
**Pull request by mattpolzin**
- Fix bug that put commas between column constraints in table schemas.
- ~Open up `SQLTableConstraint` so it can be composed with the SQLExpression for column foreign key constraints. This was nearly possible already, but I needed to be able to omit the `SQLTableConstraint`'s columns to let the column constraint define them -- the alternatives both seemed less ideal: either rewriting the column constraint for use here or relaxing column constraint to not require fields even though there is no circumstance under which the column constraint should not have fields specified.~

~I'll write some tests if the changes are approved.~

[EDIT] Per discussion below, `SQLTableConstraint` and `SQLColumnConstraint` were replaced with `SQLConstraint` and `SQLConstraintAlgorithm` (which was really aimed at column constraints in particular) was replaced by `SQLColumnConstraintAlgorithm` and `SQLTableConstraintAlgorithm`.

These changes were made to facilitate the `fluent-kit` changes here: https://github.com/vapor/fluent-kit/pull/83 but additionally set us up for representing a broader range of table constraints in general.

### CODE REVIEW

File: Sources/SQLKit/Query/SQLTableConstraint.swift [hunk 1]
```diff
@@ -1,13 +1,17 @@
 public struct SQLTableConstraint: SQLExpression {
-    public var columns: [SQLExpression]
+    /// Serialized after the name and algorithm but before the modifier
+    public var columns: [SQLExpression]?
```
Comment (tanner0101) [inline Sources/SQLKit/Query/SQLTableConstraint.swift]: I think we want to be using [`SQLColumnConstraint`](https://github.com/vapor/sql-kit/blob/master/Sources/SQLKit/Query/SQLColumnConstraint.swift) here instead.

**Comment by mattpolzin** [comment]
@tanner0101 yeah, I did not need to add any new types, I just needed to modify the `SQLTableConstraint` because its assumptions about what a table constraint looks like were not compatible with how a foreign key table constraint is formatted.

The existing `SQLTableConstraint` mandated that the last thing it serialized was a `SQLGroupExpression` of column names. Foreign key constraints must be allowed to specify "on update" and "on delete" clauses _after_ the child fields, parent table name, and parent fields.

I took an approach that should not break any existing code: Made columns optional and only serialized them if they are non-nil and added a modifier that defaults to `nil` but if specified gets serialized last. This allowed me to create a `SQLTableConstraint` that serialized the previously impossible syntax `"CONSTRAINT <name> FOREIGN KEY (<child fields>) REFERENCES <parent table> (<parent fields>) [ON UPDATE <clause>] [ON DELETE <clause>]"`.

Additionally, I chose to reuse the foreign key column constraint in producing the foreign key table constraint because after `"CONSTRAINT <name> FOREIGN KEY (<child fields>)"` the two constraint syntaxes converge.

**Comment by tanner0101** [comment]
@mattpolzin what are you plugging into `SQLTableConstraint.algorithm` then? I imagine you are plugging in `SQLConstraintAlgorithm.foreignKey(...)` into the `SQLTableConstraint.modifier` field?

While that does work, I think it's a bit tricky to understand. I wonder if this can be improved.

Column constraints take this form:

```sql
[ CONSTRAINT constraint_name ] 
    { REFERENCES reftable [ ( refcolumn ) ] 
        [ ON DELETE action ] [ ON UPDATE action ] }
```

Table constraints take this form:

```sql
[ CONSTRAINT constraint_name ]
    { FOREIGN KEY ( column_name [, ... ] ) 
        REFERENCES reftable [ ( refcolumn [, ... ] ) ] 
       [ ON DELETE action ] [ ON UPDATE action ] }
```

I see two patterns here:

```sql
[ CONSTRAINT constraint_name ] <expression>
```
Let's call this `SQLConstraint`

```sql
REFERENCES reftable [ ( refcolumn [, ... ] ) ] [ ON DELETE action ] [ ON UPDATE action ]
```
Let's call this `SQLReferences`

Now the key to getting table constraints working is putting this clause between the `SQLConstraint` and `SQLReferences`: 

```sql
FOREIGN KEY ( column_name [, ... ] ) <expression>
```

This could be an enum `SQLTableConstraintAlgorithm` with case `foreignKey` that we plug into `SQLConstraint`. This could accept a `SQLReferences` or use one internally to serialize the `REFERENCES` part of the query. 

We could then have a `SQLColumnConstraintAlgorithm` enum that also has a case `foreignKey`, but basically just directly serializes a `SQLReferences`. 

This would require a small bit of working, but I think it would be worth it. The current design of `SQLConstraintAlgorithm` doesn't really match well with SQL syntax for column / table constraints. 

Let me know if that makes sense.

**Comment by mattpolzin** [comment]
@tanner0101 I have not had time to study your response but upon skimming I am on the same page: The current setup is not quite right, and now that I know we can prioritize reworking it over changing as little as possible, I'll revisit the code.

Thanks.

**Comment by tanner0101** [comment]
@mattpolzin cool. Let me know if you have any questions. I'm happy to work on this together. Feel free to DM me on discord.gg/vapor

**Comment by mattpolzin** [comment]
@tanner0101 before I write tests I'm curious for your overall thoughts on the changes I've made. I read your thoughts and then tried to start over mentally to see where that got me and I ended up in roughly the same place you did.

Just looking for feedback on the shape of things; I'll write some tests and suss out typos and logical problems if this still feels right to you.

This is definitely a breaking change for the referenced `Fluent-Kit` PR and it is probably a breaking change in other places as well -- I will revisit that PR and make sure these changes build into the rest of the alpha Vapor packages.

File: Sources/SQLKit/Query/SQLColumnConstraintAlgorithm.swift [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    case check(SQLExpression)
+
+    /// `COLLATE` column constraint.
+    case collate(name: SQLExpression)
+
+    /// `DEFAULT` column constraint.
+    case `default`(SQLExpression)
+
+    /// `FOREIGN KEY` column constraint.
+    case foreignKey(SQLForeignKey)
```
Comment (tanner0101) [inline Sources/SQLKit/Query/SQLColumnConstraintAlgorithm.swift]: should be `SQLExpression` here

File: Sources/SQLKit/Query/SQLTableConstraintAlgorithm.swift [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    case primaryKey(columns: [SQLExpression])
+
+    /// `UNIQUE` table constraint.
+    case unique(columns: [SQLExpression])
+
+    /// `CHECK` table constraint.
+    case check(SQLExpression)
+
+    /// `FOREIGN KEY` table constraint.
+    case foreignKey(columns: [SQLExpression], SQLForeignKey)
```
Comment (tanner0101) [inline Sources/SQLKit/Query/SQLTableConstraintAlgorithm.swift]: Also `SQLExpression` here for the for `SQLForeignKey`

**Comment by mattpolzin** [comment]
Current test failure (`threw error "MySQL error: A secure connection to the server is required for authentication."`) does not appear to have to do with changes made in this PR.

File: Sources/SQLKit/Query/SQLColumnConstraintAlgorithm.swift [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    case notNull
+
+    /// `UNIQUE` column constraint.
+    case unique
+
+    /// `CHECK` column constraint.
+    case check(SQLExpression)
+
+    /// `COLLATE` column constraint.
+    case collateConstraint(name: SQLExpression)
```
Comment (tanner0101) [inline Sources/SQLKit/Query/SQLColumnConstraintAlgorithm.swift]: Why is this suffixed with `Constraint`?

File: Sources/SQLKit/Query/SQLColumnConstraintAlgorithm.swift [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+    /// `DEFAULT` column constraint.
+    case defaultConstraint(SQLExpression)
```
Comment (tanner0101) [inline Sources/SQLKit/Query/SQLColumnConstraintAlgorithm.swift]: Can we use backticks here to make this `case default` ?

Comment (mattpolzin) [reply Sources/SQLKit/Query/SQLColumnConstraintAlgorithm.swift]: This is a side-effect of the fact that a "`default`" case with one associated value conflicts with all `static` "`default`" functions of one argument. We can either have 
```
case `default`(SQLExpression)
``` 
*or*
```
case defaultConstraint(SQLExpression)
static func `default`(_ value: String)
static func `default`<T: BinaryInteger>(_ value: T)
static func `default`<T: FloatingPoint>(_ value: T)
static func `default`(_ value: Bool)
static func `default`(_ value: SQLExpression)
```

The latter provides a lot of utility (including being able to still write `.default(SQLRaw("something"))`) and the only cost is that when you are actually pattern-matching on a `SQLColumnConstraintAlgorithm` value you end up using `.defaultConstraint`.

The aforementioned (and numerous) convenience constructors come into play when adding a column to a `SQLColumnBuilder` via the `column(_: dataType: _: _:)` method.

Comment (mattpolzin) [reply Sources/SQLKit/Query/SQLColumnConstraintAlgorithm.swift]: An alternative spelling of the case that also would not collide is 
```swift
case `default`(expression: SQLExpression)
```
and I guess I don't have a preference.

**Comment by mattpolzin** [comment]
@tanner0101 sorry for the ping at what I am sure is a crazy busy time; it seems like we are _really_ close on this and the associated fluent-kit PR (https://github.com/vapor/fluent-kit/pull/83). I’d be interested in tearing off some enum work next but hoping to resolve these PRs first.

**Pull request merged**